mat64/io.go

```diff
@@ -179,3 +293,71 @@ func (v *Vector) UnmarshalBinary(data []byte) error {
 
 	return nil
 }
+
+// UnmarshalBinaryFrom decodes the binary form into the receiver, from the
+// io.Reader and returns the number of bytes read and an error if any.
+// It panics if the receiver is a non-zero Vector.
+//
+// See MarshalBinary for the on-disk layout.
+// See UnmarshalBinary for the list of sanity checks performed on the input.
+func (v *Vector) UnmarshalBinaryFrom(r io.Reader) (int, error) {
+	if !v.isZero() {
+		panic("mat64: unmarshal into non-zero vector")
+	}
+
+	var (
+		n   int
+		buf [8]byte
+	)
+	nn, err := readFull(r, buf[:])
+	n += nn
+	if err != nil {
+		return n, err
+	}
+	sz := int64(binary.LittleEndian.Uint64(buf[:]))
+	if sz < 0 {
+		return n, errBadSize
+	}
+	if sz > maxLen {
+		return n, errTooBig
+	}
+
+	v.reuseAs(int(sz))
+	for i := range v.mat.Data {
+		nn, err = readFull(r, buf[:])
+		n += nn
+		if err != nil {
+			return n, err
+		}
+		v.mat.Data[i] = math.Float64frombits(binary.LittleEndian.Uint64(buf[:]))
+	}
+
+	if n != sizeInt64+int(sz)*sizeFloat64 {
+		return n, io.ErrUnexpectedEOF
+	}
+
+	return n, nil
+}
+
+// readFull reads from r into buf until it has read len(buf).
+// It returns the number of bytes copied and an error if fewer bytes were read.
+// If an EOF happens after reading fewer than len(buf) bytes, io.ErrUnexpectedEOF is returned.
+func readFull(r io.Reader, buf []byte) (int, error) {
+	var (
+		n   int
+		max = len(buf)
+		err error
+	)
+	for n < max && err == nil {
+		var nn int
+		nn, err = r.Read(buf[n:])
+		n += nn
+	}
+	if n != max {
+		err = io.ErrUnexpectedEOF
+	}
+	if err == io.EOF {
```

Comment:
I don't believe this is correct. It is valid for an io.Reader to successfully complete a read, filling the provided byte slice and ending at the end of file to also return an io.EOF. That would mean here we get the expected number of bytes and an io.ErrUnexpectedEOF.

The correct code looks like this:

```
func readFull(r Reader, buf []byte) (int, error) {
	var n int
	var err error
	for n < len(buf) && err == nil {
		var nn int
		nn, err = r.Read(buf[n:])
		n += nn
	}
	if n == len(buf) {
		return n, nil
	}
	if n > 0 && err == io.EOF {
		return n, io.ErrUnexpectedEOF
	}
	return n, err
}
```